Core extensions for Budgets API

NobodysNightmare opened the pull request:
## OpenProject work package

https://community.openproject.org/work_packages/19141
## Description

This PR provides core side extensions to be able to make the budget in `openproject-costs` patchable.

Notably:
- allow to mount new API endpoints into existing ones
- more configurability for schema helpers

NobodysNightmare committed:
allow adding API endpoints and pathes

NobodysNightmare committed:
allow to patch existing API endpoints

NobodysNightmare committed:
actually use the magic available

NobodysNightmare committed:
allow to hide all kinds of schema

TeatroIO commented:
I've prepared a stage to preview changes. <a href='http://nobodysnightmare-feature-budgets-api.opf-openproject-2af22bf12c7dfdad4ddf.ttrcloud.com' target='_blank'>Open stage</a> or <a href='http://nobodysnightmare-feature-budgets-api.opf-openproject-2af22bf12c7dfdad4ddf.ttrcloud.com/teatro.log' target='_blank'>view logs</a>.

NobodysNightmare committed:
add method to WP schema representer

- common with other similar methods
- should consider merging of those methods some day...

NobodysNightmare committed:
use same names and magic in payload as everywhere

as introduced in Single

NobodysNightmare committed:
allow to hide link attributes from payload

NobodysNightmare committed:
style

NobodysNightmare committed:
hide spentHours on schema when it is hidden on WP

NobodysNightmare committed:
style

ulferts commented on lib/api/patchable_api.rb:
```diff
@@ -0,0 +1,71 @@
+      alias_method :group, :namespace
+      alias_method :resource, :namespace
+      alias_method :resources, :namespace
+      alias_method :segment, :namespace
```
This sounds like it might break eventually, if a new version of grape relies on different aliases. Can we do that programatically (check which alias exists) or use a different mean of overriding `namespace` (alias_method_chain)?

NobodysNightmare replied:
> This sounds like it might break eventually

That is why I put the comment-as-excuse there xD

> Can we do that programatically (check which alias exists) or use a different mean of overriding namespace (alias_method_chain)?

I'll have to google up some words you said, before answering that question. Possibly yes ^^

NobodysNightmare replied:
Okay, so `alias_method_chain` will not work, because the original `alias_method` calls all created their own **copy** of the `namespace` method. That is exactly the behaviour we would be using in an `alias_method_chain`. Thus our own redefinition and `alias_method_chain`-ing would have no effect on those original `alias_method`s.

Quick example:

``` ruby
[1] pry(main)> class Klaus
[1] pry(main)*   def foo  
[1] pry(main)*     'lame'    
[1] pry(main)*   end  
[1] pry(main)*   alias_method :bar, :foo
[1] pry(main)*   alias_method :foo_without_coolness, :foo  
[1] pry(main)*   def foo_with_coolness
[1] pry(main)*     "cool, not #{foo_without_coolness}"
[1] pry(main)*   end  
[1] pry(main)*   alias_method :foo, :foo_with_coolness
[1] pry(main)* end  
=> Klaus
[2] pry(main)> Klaus.new.foo
=> "cool, not lame"
[3] pry(main)> Klaus.new.bar
=> "lame"
```

NobodysNightmare replied:
This will work, but is quite a lot of code: http://stackoverflow.com/a/5911733/968531

ulferts commented on lib/open_project/plugins/acts_as_op_engine.rb:
```diff
@@ -141,6 +141,27 @@ def self.included(base)
+        config.to_prepare do
+          ::API::V3::Utilities::PathHelper::ApiV3Path.class_eval do
+            metaclass = class << self; self; end
+            metaclass.instance_eval do
```
There is a method for that now: http://apidock.com/ruby/Object/singleton_class

ulferts commented:
Looks good generally, will wait for amends but am willing to merge eventually.

NobodysNightmare committed:
use built-in methods instead of self-cooked ones
